pkg/storage/unified/resource/index.go

```diff
@@ -0,0 +1,194 @@
+package resource
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"log"
+	"os"
+
+	"github.com/blevesearch/bleve/v2"
+	"github.com/blevesearch/bleve/v2/mapping"
+	"github.com/google/uuid"
+)
+
+type Shard struct {
+	index bleve.Index
+	path  string
+	batch *bleve.Batch
+}
+
+type Index struct {
+	shards map[string]Shard
+	opts   Opts
+	s      *server
+}
+
+func NewIndex(s *server, opts Opts) *Index {
+	idx := &Index{
+		s:      s,
+		opts:   opts,
+		shards: make(map[string]Shard),
+	}
+	return idx
+}
+
+func (i *Index) Init(ctx context.Context) error {
+	// TODO: configure resources to index in ini file?
+	r := &ListRequest{Options: &ListOptions{
+		Key: &ResourceKey{
+			Group:    "playlist.grafana.app",
+			Resource: "playlists",
+		},
+	}}
+
+	list, err := i.s.List(ctx, r)
+	if err != nil {
+		return err
+	}
+
+	for _, obj := range list.Items {
+		res, err := getResource(obj.Value)
+		if err != nil {
+			return err
+		}
+
+		shard, err := i.getShard(tenant(res))
+		if err != nil {
+			return err
+		}
+		err = shard.batch.Index(res.Metadata.Uid, obj)
+		if err != nil {
+			return err
+		}
+	}
+
+	for _, shard := range i.shards {
+		err := shard.index.Batch(shard.batch)
+		if err != nil {
+			return err
+		}
+		shard.batch.Reset()
+	}
+
+	return nil
+}
+
+func (i *Index) Index(ctx context.Context, data *Data) error {
+	res, err := getResource(data.Value.Value)
+	if err != nil {
+		return err
+	}
+	tenant := tenant(res)
+	shard, err := i.getShard(tenant)
+	if err != nil {
+		return err
+	}
+	err = shard.index.Index(res.Metadata.Uid, data.Value.Value)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (i *Index) Delete(ctx context.Context, uid string, key *ResourceKey) error {
+	shard, err := i.getShard(key.Namespace)
+	if err != nil {
+		return err
+	}
+	err = shard.index.Delete(uid)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func tenant(res *Resource) string {
+	return res.Metadata.Namespace
+}
+
+type Metadata struct {
+	Name              string
+	Namespace         string
+	Uid               string
+	CreationTimestamp string
+	Labels            map[string]string
+	Annotations       map[string]string
+}
+
+type Resource struct {
```

Comment:
instead of having to create a resource type, we can use the [unstructured k8s library](https://pkg.go.dev/k8s.io/apimachinery@v0.31.0/pkg/apis/meta/v1/unstructured) as it has support to read and modify anything implementing a runtime.Object interface

Reply:
i think its better to have concrete types where we can.  won't all resources have these fields?  i understand the underlying json value will be dynamic, but the fields represented on these types won't change correct?